subsys/usb/class/cdc_acm.c

```diff
@@ -228,11 +232,17 @@ int cdc_acm_class_handle_req(struct usb_setup_packet *pSetup,
 	case SET_LINE_CODING:
 		memcpy(&dev_data->line_coding,
 		       *data, sizeof(dev_data->line_coding));
+		rate = sys_le32_to_cpu(dev_data->line_coding.dwDTERate);
 		LOG_DBG("CDC_SET_LINE_CODING %d %d %d %d",
-			sys_le32_to_cpu(dev_data->line_coding.dwDTERate),
+			rate,
 			dev_data->line_coding.bCharFormat,
 			dev_data->line_coding.bParityType,
 			dev_data->line_coding.bDataBits);
+#if defined(CONFIG_CDC_ACM_ARDUINO_RESET_SUPPORT)
+		if (rate == 1200 && dev_data->reset_cb != NULL) {
```

Comment:
I should be clear: I'm fine with generalizing this from detecting a change to 1200 baud to detecting a change in baud rate from the last observed (or configured) value.  But unless there's evidence this can be done on other UART platforms I believe this should be a CDC-ACM--specific extension, and I don't see any reason to generalize from baud rate change to arbitrary line control change.

Reply:
As an alternative, I've changed this to `cdc_acm_dte_rate_callback_set` that is called when the USB host changes the baud rate.
The only use case is resetting into the bootloader, but now the CDC ACM code has nothing Arduino specific in it.

Reply:
PTAL

Reply:
I actually agree with the proposed change by @nzmichaelh here. I know that for example, the J-Link CDC ACM implementation inside our debugger chips use the exact same logic (baudrate change from Host) to detect the fact that the UART has been "connected to" from the Host, i.e. that a program has opened it.

@jfischer-phytec-iot what is the argument here against this current approach?

Reply:
>It is pure generic code, and then there is this very specific to a platform(IDE?) and architecture. IMHO, we should generally not allow such a thing. Which specific IDE bootloader method is next?

No, this is actually the only reliable way a USB CDC ACM implementation can detect that a Host has opened a serial port, as far as I know.

Reply:
Note that the use case and implementation here does not detect that a host has opened a serial port; it detects that the host has opened a serial port *at a different baud rate*.  I suppose this could be converted to a "serial-port-opened" callback, if there's a reason to provide that.

I do have a use case for that sort of thing but was intending to handle it based on notifications of whether USB was connected or not, though I haven't had time to look into how that can be made to work.

Reply:
> I do have a use case for that sort of thing but was intending to handle it based on notifications of whether USB was connected or not, though I haven't had time to look into how that can be made to work.

You can use DTR for that, see samples/subsys/usb/cdc_acm

Reply:
> I actually agree with the proposed change by @nzmichaelh here. I know that for example, the J-Link CDC ACM implementation inside our debugger chips use the exact same logic (baudrate change from Host) to detect the fact that the UART has been "connected to" from the Host, i.e. that a program has opened it.

> 
> @jfischer-phytec-iot what is the argument here against this current approach?

Generally I find this approach to go into bootloader mode badly designed. It is not that there are no other methods, for example USB DFU. Here whatever code is executed in the context of USB controll request, for me it is not right, but I probably would have no problem to use simple IPC (like DBUS) for that. This code is inserted and needs to be maintained, just for the reason that some IDE can trigger the bootloader for some types of boards without pressing the button. By the way, it only works if USB is included in the firmware.
That is why I do not like this approach. I have left some comments, otherwise I leave the decision if it should go in or not to others.

Reply:
Thanks Johann.  To clarify - this isn't for a new bootloader design, but for compatibility with the manufacturer's default bootloader.  I agree if we were starting from a greenfield then something like DFU is better.

Reply:
>You can use DTR for that, see samples/subsys/usb/cdc_acm

I am not sure that works on all platforms @jfischer-phytec-iot. From experience with J-Link's CDC ACM, I have found that only a baud rate **change** is reliable in getting through on all OSs. CC @emob-nordic. Maybe that's why this BOSSA bootloader uses this approach.

Unfortunately we don't have a system-wide notification message today so that users can subscribe and listen to baud rate (or any other hardware parameters) notifications. In the meantime, I would argue that this has the potential to be useful for others and I'd vote for its inclusion.

Reply:
DTR (and RTS) over SetControlLineState is what the spec offers, SetLineCoding is misuse.

Reply:
> DTR (and RTS) over SetControlLineState is what the spec offers, SetLineCoding is misuse.

I agree it's a misuse, but it's a practical misuse, a quirk.

Reply:
Yeah, like running a red light.